src/ipnet.rs

```diff
@@ -0,0 +1,828 @@
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr & self.mask())
+    }
+
+    /// Return the network address for subnets wider than a /31 or /32.
+    pub fn network(&self) -> Option<Ipv4Addr> {
+        (self.width < 31).then(|| {
+            let addr: u32 = self.addr.into();
+            Ipv4Addr::from(addr & self.mask())
+        })
+    }
+    /// Return the broadcast address for subnets wider than a /31 or /32.
+    pub fn broadcast(&self) -> Option<Ipv4Addr> {
+        (self.width < 31).then(|| {
+            let addr: u32 = self.addr.into();
+            Ipv4Addr::from(addr | !self.mask())
+        })
+    }
+
+    /// Return the first address within this subnet.
+    pub fn first_addr(&self) -> Ipv4Addr {
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr & self.mask())
+    }
+
+    /// Return the last address within this subnet.
+    pub fn last_addr(&self) -> Ipv4Addr {
+        let addr: u32 = self.addr.into();
+        Ipv4Addr::from(addr | !self.mask())
+    }
+
+    /// Return the first host address within this subnet; for subnets wider
+    /// than a /31 or /32 that is the first address after the network address.
+    pub fn first_host(&self) -> Ipv4Addr {
+        let mask = self.mask();
+        let addr: u32 = self.addr.into();
+        let first = addr & mask;
+        if self.width == 31 || self.width == 32 {
+            Ipv4Addr::from(first)
+        } else {
+            Ipv4Addr::from(first + 1)
+        }
+    }
+
+    /// Return the last host address within this subnet; for subnets wider
+    /// than a /31 or /32 that is the address immediately before the broadcast
+    /// address.
+    pub fn last_host(&self) -> Ipv4Addr {
+        let mask = self.mask();
+        let addr: u32 = self.addr.into();
+        let last = addr | !mask;
+        if self.width == 31 || self.width == 32 {
+            Ipv4Addr::from(last)
+        } else {
+            Ipv4Addr::from(last - 1)
+        }
+    }
+
+    /// Return `true` iff the given IP address is within the subnet.
+    pub fn contains(&self, other: Ipv4Addr) -> bool {
+        let mask = self.mask();
+        let addr: u32 = self.addr.into();
+        let other: u32 = other.into();
+
+        (addr & mask) == (other & mask)
+    }
+
+    /// Produce an Interator over all addresses within this subnet.
+    pub fn addr_iter(&self) -> impl Iterator<Item = Ipv4Addr> {
+        let mask = self.mask();
+        let addr: u32 = self.addr.into();
+        let first = addr & mask;
+        let last = addr | (!mask);
+        Ipv4NetIter {
+            next: Some(first),
+            last,
+        }
+    }
+
+    /// Produce an Iterator over all hosts within this subnet. For /31 and /32
+    /// subnets, this is all addresses; for all larger subnets this excludes
+    /// the first (network) and last (broadcast) addresses.
+    pub fn host_iter(&self) -> impl Iterator<Item = Ipv4Addr> {
+        let mask = self.mask();
```

Comment:
Same comment here, I think we should delegate to `{first,last}_host()` in here. It would make very clear what we intended, and probably help testing too.